src/main/java/com/commercetools/sync/services/impl/ProductServiceImpl.java

```diff
@@ -74,35 +116,103 @@ public ProductServiceImpl(@Nonnull final ProductSyncOptions syncOptions) {
+      CompletionStage<Set<ProductProjection>> fetchMatchingResourcesInternal(
+          @Nonnull final Set<String> keys,
+          @Nonnull final Function<T, String> keyMapper,
+          @Nonnull final Function<Set<String>, Q> keysQueryMapper) {
+
+    if (keys.isEmpty()) {
+      return CompletableFuture.completedFuture(Collections.emptySet());
+    }
+
+    final List<List<String>> chunkedKeys = ChunkUtils.chunk(keys, CHUNK_SIZE);
+
+    List<Q> keysQueryMapperList =
+        chunkedKeys.stream()
+            .map(_keys -> keysQueryMapper.apply(new HashSet<>(_keys)))
+            .collect(toList());
+
+    return ChunkUtils.executeChunks(syncOptions.getCtpClient(), keysQueryMapperList)
+        .thenApply(ChunkUtils::flattenPagedQueryResults)
+        .thenApply(
+            chunk -> {
+              chunk.forEach(
+                  resource -> keyToIdCache.put(keyMapper.apply(resource), resource.getId()));
+              return new HashSet<>(chunk);
+            });
   }
 
   @Nonnull
   @Override
-  public CompletionStage<Optional<Product>> fetchProduct(@Nullable final String key) {
+  public CompletionStage<Optional<ProductProjection>> fetchProduct(@Nullable final String key) {
 
-    return fetchResource(
-        key,
-        () -> ProductQuery.of().withPredicates(buildProductKeysQueryPredicate(singleton(key))));
+    if (isBlank(key)) {
+      return CompletableFuture.completedFuture(Optional.empty());
+    }
+
+    return syncOptions
+        .getCtpClient()
+        .execute(
+            ProductProjectionQuery.ofStaged()
+                .withPredicates(buildProductKeysQueryPredicate(singleton(key))))
+        .thenApply(
+            pagedQueryResult ->
+                pagedQueryResult
+                    .head()
+                    .map(
+                        resource -> {
+                          keyToIdCache.put(key, resource.getId());
+                          return resource;
+                        }));
   }
 
   @Nonnull
   @Override
-  public CompletionStage<Optional<Product>> createProduct(
+  public CompletionStage<Optional<ProductProjection>> createProduct(
       @Nonnull final ProductDraft productDraft) {
-    return createResource(productDraft, ProductCreateCommand::of);
+    return createResource(productDraft, ProductCreateCommand::of)
+        .thenApply(product -> product.map(opt -> opt.toProjection(STAGED)));
   }
 
   @Nonnull
-  @Override
-  public CompletionStage<Product> updateProduct(
-      @Nonnull final Product product, @Nonnull final List<UpdateAction<Product>> updateActions) {
-    return updateResource(product, ProductUpdateCommand::of, updateActions);
+  public CompletionStage<ProductProjection> updateProduct(
+      @Nonnull final ProductProjection productProjection,
+      @Nonnull final List<UpdateAction<Product>> updateActions) {
+
+    return updateResourceByKey(productProjection, updateActions);
+  }
+
+  @Nonnull
+  CompletionStage<ProductProjection> updateResourceByKey(
```

Comment:
I could not see any reference for the discussed support ticket or JVM SDK issue link as a todo in this PR, could you please create and state in the code as a todo, otherwise as a  result we could easily forget why we needed this change in first place and what we need to refactor after the issue resolved ?

Reply:
887f3440d

Reply:
The issue is resolved, maybe you need to check and communicate: https://github.com/commercetools/commercetools-jvm-sdk/issues/2129#issuecomment-805272964

Reply:
Besides my comment in the JVM SDK issue, i just stumbled across your PR as it had been linked.

Mainly the issue wasn't about that you could not use update by ID, but that the ProductProjection itself was not allowed to be used in a ProductUpdateCommand.

This will be fixed with the next released SDK version. See https://github.com/commercetools/commercetools-jvm-sdk/commit/9c9c7433d0b77594b698aa8be9da6981c92e4a0b

Then a ProductProjection will also be a Versioned<Product>. If you don't want to wait there is always the workaround with Versioned.of(id, version)

Reply:
Hi @jenschude Thanks for your hint. We have already build a Workaround for this. So think we can wait, till the next release of the sdk, to remove the current workaround. Do you already know when the release will be done?

Reply:
Release will be latest April 1st 😉

Reply:
perfect, thanks for the info